adr/query_deduplication.tla

```diff
@@ -0,0 +1,501 @@
+  {
+  
+ 
+    start:
+        \* let mut locked_cache = self.cached.lock().await;
+        call lock(cache_lock);
+
+    cache_locked:
+        \*if let Some(value) = locked_cache.get(&key).cloned() { return value }
+        if (cache[Key]) {
+            \* if there is a value for the key, we test two possible behaviours
+            \* expire the key, or assume a normal task that will just return after
+            \* receiving the value
+            either {
+                \* expire one key
+                expiration_lock_1:
+                    cache[Key] := FALSE;
+                    call unlock(cache_lock);
+
+                expiration_done:
+                    goto finished;
+            }
+            or {
+                \*if let Some(value) = locked_cache.get(&key).cloned() { return value }
+                call unlock(cache_lock);
+                cache_hit:
+                    goto finished;
+            }
+        };
+
+    cache_miss:
+        \* we test at the end if all keys have been loaded
+        tested_keys := tested_keys \union {Key};
+    
+    lock_2:
+        \* let mut locked_wait_map = self.wait_map.lock().await;
+        call lock(wait_map_lock);
+    
+    lock_3:
+        \* drop(locked_cache);
+        call unlock(cache_lock);
+    
+    lookup_wait_map:
+        \* match locked_wait_map.get_mut(&key) {
+        \*    Some(waiter) => {
+        if (Cardinality(wait_map[Key]) > 0)
+        {
+            wait_map_subscribe:
+                \* we register to reeive a notification when the value is available
+                \* let mut receiver = waiter.subscribe();
+                wait_map[Key] := wait_map[Key] \union {self};
+                \* drop(locked_wait_map);
+                call unlock(wait_map_lock);
+            
+            waiting_for_value:
+                \* let (recv_key, recv_value) = receiver.recv().await
+                await channels[self];
+                goto finished;
+    
+        }
+        else
+        {
+            wait_map_register:
+                \* the wait map should be emty for this key
+                assert wait_map[Key] = {};
+                \* we should not have a value yet
+                assert ~cache[Key];
+                
+                \* let (tx, _rx) = broadcast::channel(1);
+                \* locked_wait_map.insert(key.clone(), tx.clone());
+                wait_map[Key] := {self};
+                \* drop(locked_wait_map);
+                call unlock(wait_map_lock);
+                
+            fetching_value:
+                \* let value = self.resolver.retrieve(key.clone()).await;
+                skip;
+            
+            inserting_value_lock_cache:
+                \* let mut locked_cache = self.cached.lock().await;
+                call lock(cache_lock);
+                
+            inserting_value:
+                \* locked_cache.put(key.clone(), value.clone());
+                cache[Key] := TRUE;
```

Comment:
after locking the cache, we insert the value in it (setting `TRUE`). In the original code, the value is cloned before putting into the cache and the value is sent directly to the waiting tasks, to handle eviction between the moment we put it into the cache and the moment we notify the tasks